src/Symfony/Bundle/FrameworkBundle/Client.php

```diff
@@ -104,6 +107,46 @@ public function enableReboot()
         $this->reboot = true;
     }
 
+    /**
+     * Logins a user against a specific firewall.
+     *
+     * @param string|object $user
+     * @param string[]      $roles
+     * @param string        $firewallName
+     * @param string|null   $firewallContext
+     */
+    public function login(
+        $user,
+        array $roles = array(),
+        string $firewallName = 'main',
+        string $firewallContext = null
+    ) {
+        if (!class_exists(UsernamePasswordToken::class)) {
+            throw new \RuntimeException('You must install the Symfony Security component to use this feature.');
+        }
+
+        $context = $firewallContext ?? $firewallName;
+        if (empty($roles) && $user instanceof UserInterface) {
```

Comment:
What happens when `$user` has roles, and `$roles` is also set? Currently the value of `$roles` would be completely ignored, but maybe the union of `$user->getRoles()` and `$roles` should be used, so it can be used to temporarily grant user admin privileges for example without changing their own roles?

If that is not desired, and `$roles` is just there in case `$user` is a string, I would throw an exception when `$user` is in instance of `UserInterface` and `$roles` is also set. Just to avoid any confusion as to why passing `$roles` doesn't seem to have any effect.

Reply:
Security puts the user's roles in the token when creating it, so I guess we can do the same thing + merge custom roles in the token when creating it, so it's more flexible and customizable

Reply:
IMO, I think it's a good idea to merge roles as it would be easier to play with roles and prevent from creating too many users.

Reply:
I've updated the code consequently.